files/gen_sys_stats.rb

```diff
@@ -0,0 +1,252 @@
+#!/opt/puppetlabs/puppet/bin/ruby
+# frozen_string_literal: true
+
+require "optparse"
+require "json"
+require "time"
+require "fileutils"
+
+# This script is intended to be run on a puppet infrastructure node as part of the
+# puppet_metrics_collector module.  It will generate system statistics in the file format that
+# other puppet_metrics_collector stats follow.  It will poll the system for data at a given
+# interval, and then output the average to a file once per given file interval.
+
+# General namespace for SystemMetrics module
+module SystemMetrics
+  # Main class for collecting system metrics into a json file
+  #
+  # @author Randell Pelak
+  #
+  # @attr [integer] polling_interval Time in seconds between calls to poll the system for data.
+  # @attr [integer] file_interval Time in seconds between the creation of each output file.
+  # @attr [string] metric_type cpu|memory
+  # @attr [string] metrics_dir The puppet_metrics_collector output directory.
+  # @attr [boolean] verbose Verbose output
+  # @attr [string] hostname Name of the host the metrics are from. In directory name and json file.
+  #
+  class GenerateSystemMetrics
+    #
+    # Initialize class
+    #
+    # @author Randell Pelak
+    #
+    # @param [integer] polling_interval Time in seconds between calls to poll the system for data.
+    # @param [integer] file_interval Tim in seconds between the creation of each output file.
+    # @param [string] metric_type cpu|memory
+    # @param [string] metrics_dir The puppet_metrics_collector output directory.
+    # @param [boolean] verbose Verbose output
+    #
+    # @return [void]
+    def initialize(polling_interval, file_interval, metric_type, metrics_dir, verbose = false)
+      @polling_interval = polling_interval
+      @file_interval = file_interval
+      @metric_type = metric_type
+      @metrics_dir = metrics_dir
+      @verbose = verbose
+
+      @hostname = %x[hostname].strip
+      puts "Hostname is: #{@hostname}" if @verbose
+      FileUtils.mkdir_p(metrics_dir) unless File.directory?(metrics_dir)
+    end
+
+    # Run sar to collect the raw system data
+    #
+    # @author Randell Pelak
+    #
+    # @return [string] raw output from sar
+    def run_sar
+      times_to_poll = ((@file_interval / @polling_interval) + 0.5).to_i
+      # sar inputs are polling interval and how many times to poll
+      comm_flags = " -r" if @metric_type =~ /memory/
+      comm = "/usr/bin/sar #{comm_flags} #{@polling_interval} #{times_to_poll}"
+      puts "sar command is: #{comm}" if @verbose
+      %x[#{comm}]
+    end
+
+    # Confirm the output from sar is valid
+    #
+    # @author Randell Pelak
+    #
+    # @param [string] sar_output
+    #
+    # @return [boolean]
+    def sar_output_valid?(sar_output)
+      last_line = sar_output.split("\n")[-1]
```

Comment:
Refactor suggestion:  You are performing string manipulation to get data out of the sar output in several methods.  You already have code to structure the data the`convert_sar_output_to_json` method.  I suggest breaking out the code that creates the `metrics_data` hash into its own method combined with the string manipulation. If the sar output changes in the future all code related to parsing is from string output should be encapsulated in one place.

This would make this method unnecessary (raise error during hash creation if headers != averages).  The hash can then be passed to the get_* methods.

Reply:
yeah, I didn't like the way this worked out.  Thinking about it now, I am having trouble coming up with method names that make sense.
I feel like I want to write get_sar_data_hash... have it call run_sar, and then convert...  with convert checking for errors and raising if needed. This of course eliminates "sar_output_valid?" But does that sound right? Feels like get_sar_data_hash doesn't communicate that it will actually run sar.  Guess I could go with run_sar_and_get_data_hash???

Reply:
I tried out somethings, didn't like them, and tried some more.  I think I found something that works better.  But the convert_sar_output_to_json ends up being a bit of a wall of text... but it isn't very complex so that might be okay.  Tell me what you think.